mesa/model.py

```diff
@@ -71,20 +72,24 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.running = True
         self.schedule = None
         self.current_id = 0
-        self.agents_: defaultdict[type, dict] = defaultdict(dict)
+
+        self._setup_agent_registration()
 
         self._steps: int = 0
         self._time: TimeT = 0  # the model's clock
 
+    def _setup_agent_registration(self):
+        def hack():
+            return AgentSet([], self)
+
+        self.agents_by_type: defaultdict[type, AgentSet] = defaultdict(hack)
+        self._agents = {}
```

Comment:
How often are we using this dict still? Does it remain useful?

Reply:
its essential because it is the only place that contains the hard reference to agents. If that line is removed, agents are garbage collected and will magically disappear.